Adding P4's abstract syntax to formalization doc

pataei opened the pull request:
@hackedy and @jnfoster can you please review the formalization of P4's abstract syntax (section 4)?

Keep in mind that there are some small differences between our formalization and P4 spec. I enumerate them here (by "we" refers to Petr4's implementation and the document formalization, "P4" refers to P4 spec):

1. We don't include `annotations`.
2. We don't have `match-kind` and `bit` in our data types, however, we have `match-kind` in the type of expressions.
3. We don't have the unary `+`.

In general, P4 spec tries to do some of the type checking through its grammar and we have a more general grammar.

jnfoster commented on docs/petr4spec/Petr4-spec.mdk:
```diff
@@ -142,509 +149,2979 @@ p4grammar {
+
+|     |      |     |     |
+|~~~~:|:~~~~:|:~~~~|:~~~~|
+| $\baseTyp$ | $::=$ | $\boolTyp$ | $\qquad \text{booleans}$ |
```
What's the difference between $\basetype$ and $\typ$? It's not totally clear to me from the descriptions, "Data types used in P4 programs" vs. "Types of P4 expressions". Seems like the former should be a superset of the latter, but it is not.

pataei replied:
`\typ` is the type of expression that is produced by the type system whereas `\basetyp` is the type that appears in a P4 expression.

Connecting it to Petr4 implementation, `\typ` is the type used at the IR level and `\basetyp` is the type used at the surface level.

pataei replied:
break up `\typ` into different syntactic categories. 
X = type variables
x = names assigned to types declared in a program
1) surface level
2) declared types that cannot be referred 

drop W. add wildcard to parameter. drop k. 
check function type, you need parameter directions and types. 
think about including extern methods within the type. 
distinguish between records/...

pataei replied:
- [ ] in syntax, I noted that a type variable can also be a wildcard. Does the syntax need to represent this? we can have:
`X ::= x | _ `

jnfoster commented on docs/petr4spec/Petr4-spec.mdk:
```diff
@@ -142,509 +149,2979 @@ p4grammar {
+
+|     |      |     |     |
+|~~~~:|:~~~~:|:~~~~|:~~~~|
+| $\typ$ | $::=$ | $\boolTyp$ | $\qquad \text{booleans}$ |
```
Pedantic, but where there are types in common it'd be nice for them to be listed in the same order as the grammar above.

pataei replied:
@pataei apply this in your last review of the syntax.

jnfoster commented on docs/petr4spec/Petr4-spec.mdk:
```diff
@@ -142,509 +149,2979 @@ p4grammar {
+|   | $\vert$ | $\newTypeTyp \typVar \typ$ | $\qquad \text{reference}$ |
+|   | $\vert$ | $\headerTyps \field \typ$ | $\qquad \text{headers}$ |
+|   | $\vert$ | $\headerUnionTyps \field \typ $ | $\qquad \text{header unions}$ |
+|   | $\vert$ | $\structTyps \field \typ$ | $\qquad \text{structs}$ |
```
Headers, header unions, and structs are not distinguished syntactically in the grammar. This seems like a bug. Don't we need to distinguish them?

pataei replied:
Actually, as I was reviewing the typing rules of well-formed types I only kept one of them, that is, the syntax only has one record type that represents all headers, header unions, structs, and records. 

Do we want the abstract syntax to distinguish them syntactically if they all have the same structure? P4's concrete syntax distinguishes them but I don't know what its abstract syntax does.

jnfoster commented on docs/petr4spec/Petr4-spec.mdk:
```diff
@@ -142,509 +149,2979 @@ p4grammar {
+|     |      |     |     |
+|~~~~:|:~~~~:|:~~~~|:~~~~|
+| $\prmDef$ | $::=$ | $\dir \, \var : \baseTyp$ | $\qquad \text{parameters}$ |
+|   | $\vert$ | $\dir \,\var : (\baseTyp, \exp)$ | $\qquad \text{parameters with value}$ |
```
Syntax here also seems a bit convoluted. Why the parentheses around the type and expression? Again, should we just use P4's concrete syntax as our guide?

pataei replied:
applied.

jnfoster commented on docs/petr4spec/Petr4-spec.mdk:
```diff
@@ -142,509 +149,2979 @@ p4grammar {
+|~~~~:|:~~~~:|:~~~~|:~~~~|
+| $\arg$ | $::=$ | $\argexp$ | $\qquad \text{expression}$ |
+|   | $\vert$ | $\argkv$ | $\qquad \text{key value}$ |
+|   | $\vert$ | $\missingarg$ | $\qquad \text{missing}$ |
```
Here $\missingarg$ renders as "_" whereas previous "don't care" types and directionless parameters were written out in the abstract syntax. Let's be consistent?

pataei replied:
applied.

pataei replied:
- [] discuss in meeting: arguments in P4 can also be `x = dontcare` but Petr4 doesn't allow this. Include it in abstract syntax?

pataei replied:
@pataei double check this and add it to issues, if applicable.

jnfoster commented on docs/petr4spec/Petr4-spec.mdk:
```diff
@@ -142,509 +149,2979 @@ p4grammar {
+|   | $\vert$ | $\str$ | $\qquad \text{strings}$ |
+|   | $\vert$ | $\int$ | $\qquad \text{integers}$ |
+|   | $\vert$ | $n_w$ | $\qquad \text{fixed-width signed integers}$ |
+|   | $\vert$ | $b_w$ | $\qquad \text{fixed-width unsigned integers}$ |
```
Ditto.

jnfoster commented on docs/petr4spec/Petr4-spec.mdk:
```diff
@@ -142,509 +149,2979 @@ p4grammar {
+|~~~~:|:~~~~:|:~~~~|:~~~~|
+| $\nxt$ | $::=$ | $\str$ | $\qquad \text{next state}$ |
+| $\pmatch$ | $::=$ | $\defMatch$ | $\qquad \text{default match}$ |
+|   | $\vert$ | $\dontcareMatch$ | $\qquad \text{don't care match}$ |
```
Same comment about how $\dontcareMatch$ is rendered as above.

pataei replied:
- [ ] we want it to render to `dontcare`?

jnfoster commented on docs/petr4spec/Petr4-spec.mdk:
```diff
@@ -142,509 +149,2979 @@ p4grammar {
+|   | $\vert$ | $\dontcareMatch$ | $\qquad \text{don't care match}$ |
+|   | $\vert$ | $\exp$ | $\qquad \text{expression match}$ |
+| $\case$ | $::=$ | $\expandCase \pmatches \nxt$ | $\qquad \text{match cases for the next state}$ |
+| $\parserTran$ | $::=$ | $\nxt$ | $\qquad \text{next state transition}$ |
```
Same comment about echoing P4's concrete syntax when possible. Would it be more readable to have "transition" somewhere here, for example?

pataei replied:
@pataei revise parser declaration syntax.

jnfoster commented on docs/petr4spec/Petr4-spec.mdk:
```diff
@@ -142,509 +149,2979 @@ p4grammar {
+|   | $\vert$ | $\typdefDcl \typVar \dcl$ | $\qquad \text{type definitions (declarations)}$ |
+|   | $\vert$ | $\newtypeDcl \typVar \baseTyp$ | $\qquad \text{generative type definitions}$ |
+|   | $\vert$ | $\newtypeDcl \typVar \dcl$ | $\qquad \text{generative type definitions (declarations)}$ |
+|   | $\vert$ | $\controlTypDcl \typVar \typVars \prmDefs$ | $\qquad \text{control types}$ |
```
Not a big fan of the abbreviations here and below, especially as we wrote them out above. Let's be consistent?

pataei replied:
applied. I wanted to differentiate between an object declaration and a type declaration but we can stay consistent with P4 (P4 doesn't differentiate between the two).

pataei commented on docs/petr4spec/Petr4-spec.mdk:
```diff
@@ -142,509 +149,2979 @@ p4grammar {
+
+|     |      |     |     |
+|~~~~:|:~~~~:|:~~~~|:~~~~|
+| $\prmDef$ | $::=$ | $\dir \, \var : \baseTyp$ | $\qquad \text{parameters}$ |
```
@pataei apply in your revision.

pataei commented on docs/petr4spec/Petr4-spec.mdk:
```diff
@@ -142,509 +149,2979 @@ p4grammar {
+
+|     |      |     |     |
+|~~~~:|:~~~~:|:~~~~|:~~~~|
+| $\switchCase$ | $::=$ | $\actionCase \stmts$ | $\qquad \text{labeled block}$ |
```
applied.

pataei commented on docs/petr4spec/Petr4-spec.mdk:
```diff
@@ -142,509 +149,2979 @@ p4grammar {
+|   | $\vert$ | $\externFuncDcl \baseTyp \var \typVars \prmDefs$ | $\qquad \text{extern functions}$ |
+|   | $\vert$ | $\varDclInit \baseTyp \var \exp$ | $\qquad \text{local variables (initialized)}$ |
+|   | $\vert$ | $\varDcl \baseTyp \var$ | $\qquad \text{local variables (uninitialized)}$ |
+|   | $\vert$ | $\valSetDcl \baseTyp \exp \var$ | $\qquad \text{parser value sets}$ |
```
- [ ] disucss in meeting.
This might be a bit more general than P4's definition but not by much. Here's P4's definition:

```
valueSetDeclaration
  : optAnnotations
      VALUESET '<' baseType '>' '(' expression ')' name ';'
  | optAnnotations
      VALUESET '<' tupleType '>' '(' expression ')' name ';'
  | optAnnotations
      VALUESET '<' typeName '>' '(' expression ')' name ';'
  ;
```

pataei commented:
@jnfoster @hackedy section 5.1 (type well-formedness) is ready to review. 

Changes applied in the commit _4a14625_:

- Merged tuple and list types into one.
- Removed don’t care type from the syntax of types of expressions
- Added enum without underlying type to syntax
- Changed `\dir \var : \typ` to `\prm` in expression types part of the syntax
- Revised `\prm` to `\prm \typKind` where type kind can either be `\tau` or `\rho`. 
- Combined all record, header, header union, and struct types into one.
- Fixed env explanation in all rules of well-formed 

Need to discuss the followings:
- [ ] should we mention type saturation?
- [ ] Petr4 doesn't check the length of a fixed-length integer, either signed or unsigned. However, P4 spec specifies that for signed integers the length must be more than 1 and for unsigned integers the length must be equal or more than zero. We should revise Petr4's rules accordingly, all agree?
- [ ] In enum well-formedness rule: Petr4 doesn't allow for an enum to have default values (initializer) for constants, however, P4 spec does. Here's an [example](https://p4.org/p4-spec/docs/P4-16-v1.2.0.html#sec-enum-types). The current type well-formedness rule of enum also ignores the initializer. 
- [ ] In control type well-formedness rule: P4 spec states that "The types parser, control, and package cannot be used as types of arguments for methods, parsers, controls, tables, actions. They can be used as types for the arguments passed to constructors." The formalization doesn't take this into account neither does Petr4's implementation. We could check this in the well-formedness of types or even when we're declaring a method, parser, control, table, or action, instead of when we're defining their type. [p4 spec ref](https://p4.org/p4-spec/docs/P4-16-v1.2.0.html#sec-parser-control-types)

pataei commented:
- [ ] @pataei reminder: change the order of the rules to follow the order of syntax.

hackedy commented on docs/petr4spec/Petr4-spec.mdk:
```diff
@@ -142,509 +149,3025 @@ p4grammar {
+|   | $\vert$ | $\mod$ | $\qquad \text{numerical modulo}$ |
+|   | $\vert$ | $==$ | $\qquad \text{equality check}$ |
+|   | $\vert$ | $!=$ | $\qquad \text{inequality check}$ |
+|   | $\vert$ | $\plusSat$ | $\qquad \text{saturating addition}$ |
```
These operators might look better in \mathsf or \textt.

pataei replied:
It definitely looks nicer however it doesn't work with $\sim$, $\leq$, $\geq$, $\gg$, $\ll$, and $\hat{}$. Figure a way around it when it comes to review and clean up.

hackedy replied:
Ooh, that's annoying. I think `|+|` and other saturating operations are probably best rendered either as code font `|+|` or as some kind of TeX glyph like \boxplus.

Sorry if this is sending you emails btw I know it's 11 o clock.

pataei replied:
I'll try \boxplus. and no worries, I appreciate your feedback any time of the day ;)

pataei replied:
figured it out.

hackedy commented on docs/petr4spec/Petr4-spec.mdk:
```diff
@@ -142,509 +149,3025 @@ p4grammar {
+
+|     |      |     |     |
+|~~~~:|:~~~~:|:~~~~|:~~~~|
+| $\nxt$ | $::=$ | $\str$ | $\qquad \text{next state}$ |
```
Should this be nxt ::= s | accept | reject?

hackedy commented on docs/petr4spec/Petr4-spec.mdk:
```diff
@@ -142,509 +149,3025 @@ p4grammar {
+| $\prop$ | $::=$ | $\key \exp \var$ | $\qquad \text{keys}$ |
+|   | $\vert$ | $\action$ | $\qquad \text{actions}$ |
+|   | $\vert$ | $\entry \pmatch \action$ | $\qquad \text{entries}$ |
+|   | $\vert$ | $\custom \bool \var \exp$ | $\qquad \text{custom property}$ |
```
what is the \bool here for?

pataei replied:
I think it determines if the custom property is constant or not. This was the version based on Petr4 and it's defined as (I may end up dropping it after revision to base the formalization on P4):

`  type 'a pproperty =
      Key of
        { tags: 'a;
          keys: key list }
      (** Keys *)
    | Actions of
        { tags: 'a;
          actions: action_ref list }
      (** Actions *)
    | Entries of
        { tags: 'a;
          entries: entry list }
      (** Entries *)
    | Custom of
        { tags: 'a;
          annotations: Annotation.t list;
          const: bool;
          name: P4String.t;
          value: Expression.t }
      (** Custom (e.g., [size], [default_action], etc.) *)
  [@@deriving sexp,show,yojson]
`

hackedy replied:
I guess an alternative to `custom b x e` would be to have two variants `custom x e` and `custom const x e`, or something like that.

hackedy commented on docs/petr4spec/Petr4-spec.mdk:
```diff
@@ -142,509 +149,3025 @@ p4grammar {
+|   | $\vert$ | $\errDcl \fields$ | $\qquad \text{error types}$ |
+|   | $\vert$ | $\matchkindDcl \fields$ | $\qquad \text{match kind types}$ |
+|   | $\vert$ | $\enumDcl \typVar \fields$ | $\qquad \text{enumerated types}$ |
+|   | $\vert$ | $\serEnumDcl \baseTyp \typVar \field \exp$ | $\qquad \text{serializable enumerated types}$ |
```
The \field macro here rendered as f: exp, should be something like f = exp instead to indicate it's defining the value of the enum constant f

pataei replied:
good catch!

hackedy commented on docs/petr4spec/Petr4-spec.mdk:
```diff
@@ -142,509 +149,3025 @@ p4grammar {
+|   | $\vert$ | $\typdefDcl \typVar \baseTyp$ | $\qquad \text{type definitions}$ |
+|   | $\vert$ | $\typdefDcl \typVar \dcl$ | $\qquad \text{type definitions (declarations)}$ |
+|   | $\vert$ | $\newtypeDcl \typVar \baseTyp$ | $\qquad \text{generative type definitions}$ |
+|   | $\vert$ | $\newtypeDcl \typVar \dcl$ | $\qquad \text{generative type definitions (declarations)}$ |
```
IMO render newtype as [newtype X \rho] or for consistency with P4's concrete syntax [type X \rho]. The generative one shouldn't be an equation since X != \rho in the type system.

pataei replied:
shouldn't it be `type \rho t`? where `t` is a type identifier and not a type variable.

hackedy replied:
Ah, yes. I always make this mistake.

hackedy reviewed (approved):
Hi, looks great. I think this is mergeable. I left a few comments about notation.

hackedy commented:
🎊 🎊 🪂
